kernel/src/tbfheader.rs

```diff
@@ -9,8 +12,72 @@ macro_rules! align4 {
     };
 }
 
+/// Error when parsing an app's TBF header.
+pub enum TbfParseError {
+    /// Not enough bytes in the buffer to parse the expected field.
+    NotEnoughFlash,
+
+    /// The length fields of the header and app do not make sense. Likely the
+    /// TBF header length is set to be longer than the entire app.
+    BadSize,
+
+    /// Unknown version of the TBF header.
+    UnsupportedVersion(u16),
+
+    /// Checksum calculation did not match what is stored in the TBF header.
+    /// First value is the checksum provided, second value is the checksum we
+    /// calculated.
+    ChecksumMismatch(u32, u32),
+
+    /// One of the TLV entries did not parse correctly. This could happen if the
+    /// TLV.length does not match the size of a fixed-length entry. The `usize`
+    /// is the value of the "tipe" field.
+    BadTlvEntry(usize),
+
+    /// The app name in the TBF header could not be successfully parsed as a
+    /// UTF-8 string.
+    BadProcessName,
+
+    /// Internal kernel error. This is a bug inside of this library. Likely this
+    /// means that for some reason a slice was not sized properly for parsing a
+    /// certain type, which is something completely controlled by this library.
+    /// If the slice passed in is not long enough, then a `get()` call will
+    /// fail and that will trigger a different error.
+    InternalError,
+}
+
+impl From<core::array::TryFromSliceError> for TbfParseError {
+    // Convert a slice to a parsed type. Since we control how long we make our
+    // slices, this conversion should never fail. If it does, then this is a bug
+    // in this library that must be fixed.
+    fn from(_error: core::array::TryFromSliceError) -> Self {
+        TbfParseError::InternalError
+    }
+}
+
+impl fmt::Debug for TbfParseError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            TbfParseError::NotEnoughFlash => write!(f, "Buffer too short to parse TBF header"),
+            TbfParseError::BadSize => write!(f, "Invalid TBF header and app lengths"),
+            TbfParseError::UnsupportedVersion(version) => {
+                write!(f, "TBF version {} unsupported", version)
+            }
+            TbfParseError::ChecksumMismatch(app, calc) => write!(
+                f,
+                "Checksum verification failed: app:{:#x}, calc:{:#x}",
+                app, calc
+            ),
+            TbfParseError::BadTlvEntry(tipe) => write!(f, "TLV entry type {} is invalid", tipe),
+            TbfParseError::BadProcessName => write!(f, "Process name not UTF-8"),
+            TbfParseError::InternalError => write!(f, "Internal kernel error. This is a bug."),
+        }
+    }
+}
+
+// TBF structure
+
 /// TBF fields that must be present in all v2 headers.
-#[repr(C)]
```

Comment:
Quick experiment: adding `repr(c)` back did not change the resulting binary size for hail. It did make the binary larger for opentitan. So it might have some effect/benefit on RISC-V platforms.